### PR TITLE
Mech code refactor

### LINKED ISSUE #6812: MMIs in mechs cannot use the equipment or connect to chargers.
MMIs in mechs cannot use the equipment or connect to chargers. Trying to connect to a charger just gives the "Nothing happens" message, you can raise and lower equipment but can't seem to use it at all.

### LINKED ISSUE #7354: Aliens cannot chestburst from a human that's in a mech
bug was found with someone in a durand, so it might only affect the durand, or all mechs. one or the other.

### LINKED ISSUE #8110: High-pressure unactivated turfs
Apparently it's possible to make turfs that have a very high pressure (4500 kPa and more) but that aren't activated, so you have a single very high-pressure turf that doesn't spread.

```
DEAD: Quinton Bould says, "It's from a destroyed ripley that had its internal pressure set to the canister's max"
DEAD: Quinton Bould says, "so the air on the spot the ripley was destroyed"
DEAD: Quinton Bould says, "is stuck at"
DEAD: Quinton Bould says, "What the output pressure was"
```

mechs a shit, once again.

### LINKED ISSUE #9425: Mechs do not lose charge from moving, turning, or having lights on.
What it says on the tin. They still lose power if they use their modules, however.

### LINKED ISSUE #9462: Sometimes the Mecha verbs are not assigned to the pilot.
It can happen that when you enter a mech you are not given the control verbs for it, the Exosuit tab doesn't appear and you can't manually use the verbs, leaving you stuck inside the mech. Everything else works (using mech equipment, moving and the skin changes, all normal), but the verbs are not given to you.

How to reproduce: jump in a mech and hope for the worst. It's seemingly random.

### LINKED ISSUE #10330: Teleportation
When i was in a mech in my first abductor round i was stormed around by people so i clicked emergency beam and got teleported back,i was taking messsages of mech getting damaged and then when it blew,i automatically got teleported back to the dead mech and got killed,please fix it.

### LINKED ISSUE #10387: Mechs and HUD alerts
If you're in a mech, alerts for fire, temperature, and toxins in air don't appear. You just see your health decreasing if you're taking damage.

### PR DESCRIPTION
**Pull request by phil235**
- Mech equipment was using its unique processing method, with specific timing for each equipment, but now it uses obj processing, I changed the values so the equipment effect stays the same. Removed global_iterator datum.
- removed mecha/proc/do_after and /enter_after() procs. It now uses the original do_after().
- Renamed mech sleeper occupant var to "patient" to avoid confusion with the occupant of the mech.
- all non instant tool (drill) now show a progbar when drilling etc..
- action cooldown now use do_after_cooldown() (that itself uses do_after) and start_cooldown (for instantaneous actions).
- Removed mecha_equipment's destroy proc, it's now all in Destroy(). No confusion.
- modified mecha_equipment/proc/can_attach() to not check istype(mecha) b/c it can't not be. (so the child procsonly have one istype check)
- Removed diamonddrill/can_attach() , all done at drill level.
- armor booster now available to all mechs.
- Removed dynhitby, dynbulletdamage, dynattackby, dynusepower(), dyngetcharge(), dynabsorbdamage()
- I split the tools.dm file into smaller ones: work tools, mining tools, other tools.
- I split mecha.dm into mecha.dm, mecha_topic.dm and mecha_defense.dm
- refactored mech weapon ballistic/launcher, new proj_init proc, more OOP.
- Moving consumes energy! Lights consumes energy. Fixes #9425.
- Fixed #7354 xeno not bursting if host is inside a mech.
- Added action buttons to mech. Moved toggle lights, internal tank, eject, view stats from verbs to action buttons, these can now only be done via these buttons (removed them from the big stats window).
- Removed port connect verbs b/C it's automated now.
- regular hud is no longer hidden when inside a mech (doesn't matter b/c you can't interact with most stuff in it while in a mecha). Fixes #10387
- can't walk when shooting projectiles. Makes walking on your projectile a bit harder. Helps against issue #10315 (but doesn't fix it).
- also made into action buttons: the special abilities of certain combat mechs.
- Added thrown alerts for mech charge and integrity.
- Fixes teleporting occupant out not properly removing it from the mech. Fixes #10330
- fixes ballistic mech weapons spinning when launched. proc/throw_at() now has a spin argument instead of using var/allow_spin that was added to all atoms just for that.
- added a update_action_buttons() to ai/life() to handle ai mech.
- /obj/item/mecha_parts/mecha_equipment/tool/... is now just mecha_equipment/... (the /tool/ was useless)
- Fixes seraph smoke ability not working.
- Fixes #6812 (mmi in mech cant use equipment).
- Moved check to prevent drilling inventory to mecha/click_action so that it also happens with an AI occupant.
- Fixes #9462 (verbs sometimes aren't given to the pilot) by removing the verbs
- Upon destroy() we delete cabin_air: Fixes #8110 
- Fixes heavy plasma cutter not being attachable to anything.

The action buttons: eject (only for human pilot), toggle air tank usage, cycle equipment, toggle lights, view stats. The special abilities of certain mechs are also action buttons (smoke, phasing, defense mode, zoom, thrusters, switch damagetype)

### CODE REVIEW

**Comment by Anonus** [comment]
Make sure you can't drill items you're holding while you're in a mech, that's why the interface was hidden in the first place.

**Comment by GunHog** [comment]
Could you add a fix for https://github.com/tgstation/-tg-station/issues/10430 here as well?

**Comment by phil235** [comment]
@Anonus  I checked, you can't drill your inventory stuff

**Comment by WJohn** [comment]
Oh. This is why you were gone for the past week

I forgot, phil either spends a week quickly fixing 30 bugs

or completely overhauls a gigantic system no one has wanted to touch for years.

I'm at a loss for words how you can pull this off again and again <3 <3 <3

**Comment by Killbasa** [comment]
Yay

**Comment by xxalpha** [comment]
The mech tag was short lived.

:+1:

**Comment by Miauw62** [comment]
phil is code jesus

File: data/mode.txt [hunk 1]
```diff
@@ -1 +1 @@
-extended
\ No newline at end of file
+extended
```
Comment (Fox-McCloud) [inline data/mode.txt]: Uh, oops?

Comment (phil235) [inline data/mode.txt]: oh oh...

**Comment by MrStonedOne** [comment]
the eject action button needs a better button.

**Comment by phil235** [comment]
@MrStonedOne Oh yeah, the action button icons are pretty much placeholders, if someone wants to make better ones I'll happily use those instead.

**Comment by GunHog** [comment]
I am going to yoink the lighting button for my Cyborg lamps, if it fits the theme.

**Comment by optimumtact** [comment]
@GunHog I'm glad you commented, I was just about to start work on inbuilt borg lamps

**Comment by GunHog** [comment]
I already PR'ed it, silly fruit!

**Comment by Lumipharon** [comment]
Why remove the armor boosters from the ripley? It has actually reason to use them (occasionally), and has the slots to spare.

Isn't stopping mechs from moving while shooting going to make them super clunky, especially with the burst fire AC2?

**Comment by phil235** [comment]
>Why remove the armor boosters from the ripley? It has actually reason to use them (occasionally), and has the slots to spare.

Well, I only changed it because I found it weird to have the item available to all mechs except honkers, so I changed it to combat only. I'm totally fine with giving it to all mechs tbh.

>Isn't stopping mechs from moving while shooting going to make them super clunky, especially with the burst fire AC2?

Well, it barely stops them actually, the reason I said it didn't fix the issue with walking on your projectile is because the stop is barely noticeable and still do not entirely prevent walking on your bullets.

**Comment by PKPenguin321** [comment]
>Fixes #9562 (verbs sometimes aren't given to the pilot) by removing the verbs

I think you cited the wrong number, that one leads to aran's newscaster code cleanup.

**Comment by freerealestate** [comment]
:+1: Holy shit

**Comment by WJohn** [comment]
What happens if the larva bursts inside the mech? Are they deposited and the seat is opened? Are they trapped inside the mech? Do they comically gain control?

**Comment by phil235** [comment]
>I think you cited the wrong number, that one leads to aran's newscaster code cleanup.

fixed.

>What happens if the larva bursts inside the mech? Are they deposited and the seat is opened? Are they trapped inside the mech? Do they comically gain control?

The larva appears outside the mech, the human pilot is automatically ejected upon its death/gibbing.

**Comment by LaharlMontogmmery** [comment]
Holee crap, Phil.

You're actually making this happpening.

Also, since now that Mechs are not using that special Mech Equipment proc, does that mean we can now actually add equipement to mechs via VV just by adding them to the mech? That would be great. I mean, apart from OTHER things..

**Comment by GunHog** [comment]
>Second question : last time I was a mech AI (an Odysseus, in that case), I couldn't put people in the mounted sleeper. It was trying to do the whole "Follow the mob with cameras" gig instead, Tried to save the chaplain from the cult by putting him in the sleeper, didn't work. Does the modifications to click_action fixes that?

Use middle mouse button to activate mech functions. I know it is not ideal, but checking for a mech each click for the AI was considered unacceptable, so I had to change it. There is text shown to the AI upon entering to notify the player of this.

**Comment by LaharlMontogmmery** [comment]
Ho, okay, didn't know/see that. My bad.

**Comment by phil235** [comment]
>Also, I'm going to suppose that update_action_buttons allows you to have the intent UI, which is a plus for Mech AIs.

Nope, the intent UI comes from the human occupant, it unfortunately doesn't appear for AI or MMI (#10430).

**Comment by phil235** [comment]
Ready to be merged now.

**Pull request merged**